### PR TITLE
Allows ABAC authorizor to reload policy file when the file is changed

### PR DESCRIPTION
**Pull request by huang195**
Refers to #22988 and #23095

### CODE REVIEW

**Comment by huang195** [comment]
@liggitt Per your suggestion on the earlier PR. This one allows a policy file to be reloaded when it's changed, so apiserver doesn't have to be restarted for the change to take into effect. I've tested it, and the policy does get refreshed whenever the underlying file is changed. I've made 2 commits, one for the fsnotify package that needs go into Godeps, and one for the code change related to this PR, as it was suggested to do it this way for easier reviewing.

File: pkg/auth/authorizer/abac/abac.go [hunk 1]
```diff
@@ -52,7 +53,46 @@ func (p policyLoadError) Error() string {
 type policyList []*api.Policy
 
 // TODO: Have policies be created via an API call and stored in REST storage.
-func NewFromFile(path string) (policyList, error) {
+func NewFromFile(path string) (*policyList, error) {
```
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: make watching optional

File: pkg/auth/authorizer/abac/abac.go [hunk 2]
```diff
@@ -52,7 +53,46 @@ func (p policyLoadError) Error() string {
...
+	}
+
+	// Watch the policy file
+	err = watcher.Add(path)
+	if err != nil {
+		return nil, err
+	}
+
+	// On write operations on the policy file, reload policy file
+	go func() {
```
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: use `wait.Forever(func(){...}, time.Second)`, which will handle panics inside the spawned gofunc and prevent hotlooping

File: pkg/auth/authorizer/abac/abac.go [hunk 3]
```diff
@@ -52,7 +53,46 @@ func (p policyLoadError) Error() string {
...
+		for {
+			select {
+			case ev := <-watcher.Events:
+				switch ev.Op {
+				case fsnotify.Write:
```
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: I'm not super familiar with fsnotify... does it handle these cases:
1. deleting and recreating the file?
2. renaming the original file and replacing it with another file?

File: pkg/auth/authorizer/abac/abac.go [hunk 4]
```diff
@@ -52,7 +53,46 @@ func (p policyLoadError) Error() string {
...
+					_, err = newFromFile(path, pl)
```
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: load, handle errors, then swap rules... don't overwrite rules until we're sure we loaded the file correctly

Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: might need to use atomic.Value.Store/Load to do this

File: pkg/auth/authorizer/abac/abac.go (hunk 3 above)
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: also, will we get multiple events as the file is written, and could we get called when the file is half-written?

File: pkg/auth/authorizer/abac/abac.go (hunk 1 above)
Comment (huang195) [inline pkg/auth/authorizer/abac/abac.go]: Should there be a new flag defined to set this parameter

File: pkg/auth/authorizer/abac/abac.go (hunk 3 above)
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: seems like we should also watch Create events

Comment (huang195) [inline pkg/auth/authorizer/abac/abac.go]: if a file is deleted, it's no longer watched. It uses inotify under the cover, so this is a limitation of inotify. Renaming is essentially a copy+delete. Probably the only way around this problem is by polling, which is kind of ugly. If we're adding a new flag per above comment, perhaps this limitation can be clearly stated there.

File: pkg/auth/authorizer/abac/abac.go (hunk 4 above)
Comment (huang195) [inline pkg/auth/authorizer/abac/abac.go]: yep, make sense.

File: pkg/auth/authorizer/abac/abac.go (hunk 3 above)
Comment (huang195) [inline pkg/auth/authorizer/abac/abac.go]: let me investigate this a bit more to see if there's a more graceful way of handling these.

Comment (huang195) [inline pkg/auth/authorizer/abac/abac.go]: Here are the operations fsnotify is able to monitor:
*CREATE: not relevant to us as this is watching for file/dir creation within a directory. We are watching one file only.
*WRITE: it is possible that multiple writes could happen consecutively, and we should only trigger a refresh on the last write
*REMOVE: stop watching the file and inode is gone
*RENAME: as inode of the file stays the same after a `mv`, it has no consequence for us. 
*CHMOD: not relevant to us

Some text editors like vim or emacs does funny things with files, e.g., When you do a `:wq` in vim, it renames the original file, writes a new file, and then deletes the original file. Emacs does similar things with its backup file prefixed with `~`. Besides periodic flushes from a text editor could cause a file to be written when it's half written, and a reload of the file could result in unintended behaviors.

The clean way this would work is if a copy of the policy file is obtained first, and then the copy is edited, and once done, it is finally copied over the original policy file. We need to make sure if this causes multiple write operations, only the last write triggers the file reload.

**Comment by huang195** [comment]
@liggitt This is actually much messier to handle than watching for a change in a policy file stored in a REST storage, in which changes can be done atomically via SET. Watching for a file change, on the other hand, we need to deal with text editor flushing out half-written content and other idiosyncrasies of text editors, with a `mv-in`, in which the inode of the file being watched changes, and a `mv-out`, in which the inode doesn't change. There might be other corner cases to consider as well.

I can see 2 approaches that might work:
1. Make certain assumptions on how the policy file can be modified, e.g., if this is done manually, one would need to i) make a copy of the current policy file, ii) make changes in the copy, and iii) copy back the changes. And one should never delete the current policy file, or future changes will not be tracked. One should not rename (`mv`) the current policy file, as that is confusing. One should not overwrite (`mv`) the current policy file with another file as inode would change. To enforce all these rules, it is probably best to write a small script to handle changes to the policy file.
2. Revisit #23095 which allows policy file to be stored as a kv pair in an etcd and reload it when changed. We discussed whether or not it makes sense to put this file in the same etcd cluster that K8s uses or a secondary etcd cluster. Now I think about it, the amount of additional load of putting the policy file into the primary etcd cluster is very small. The only time we'll be putting additional load is when the policy is changed, but I don't expect user add/delete/update operations to be done very frequently. Moreover, for K8s running in HA mode, podmaster actually makes use of an etcd cluster (can be the primary etcd cluster) for leadership election, and it puts a much higher load to the etcd cluster because it has to constantly refresh TTL of particular keys. Can we do something similar that we give users an option to specify whether or not to load the policy from a file or etcd cluster, and in the latter case, whether or not to use the primary etcd cluster or a secondary one.

**Comment by huang195** [comment]
Maybe as a third option (probably the most clean), write an etcd-based ABAC authorizor as a plugin.

**Comment by huang195** [comment]
@liggitt @erictune OK coming back to this one. The idea is to add a restriction on how a local file can be modified so we get more consistent result when monitoring it using fsnotify. Specifically, the restriction is the policy file needs to be in a read only mode, i.e., `chmod 0444 <policy file>`. This will result in the following behavior:

- This will thwart admin from directly editing the file, resulting in having to deal with a half-written file that is not meant to be loaded into memory.
- This will thwart the policy file from being deleted
- This will thwart admin from overwriting is using `mv <another file> <policy file>` or `cp <another file> <policy file>`

Of course, an admin can override this behavior, but he needs to explicitly specify the intent, which makes this less error prone. The procedure to modify an existing policy file is as follows:

1. `cp <policy file> <new policy file>`
2. modifies <new policy file>
3. `chmod 0644 <policy file>; cp <new policy file> <policy file>; chmod 0444 <policy file>`

This will result in us getting notified by fsnotify some CHMOD and WRITE events, for which we will use to trigger a reload action.

**Comment by liggitt** [comment]
how are partial reads handled with the static pod manifest the node reads? does it just poll and hope it doesn't read in the middle of a write?

File: pkg/auth/authorizer/abac/abac.go (hunk 3 above)
Comment (erictune) [inline pkg/auth/authorizer/abac/abac.go]: If you make a copy, edit it, and then [rename](http://man7.org/linux/man-pages/man2/rename.2.html) using the rename syscall (mv should work too, I think), then the new file will appear atomically.  

You would need to document to users this requirement.

If you are polling, then this works just the way you want it to: the apiserver never sees a partial file.  

I don't see fsnotify helping you much here, versus polling every few seconds (assuming policy files are reasonably sized).

You still need to handle the case of an incorrect file.  Should you crash, as is currently the case when an incorrect file is provided?  Or keep using the old one?  How to surface this error to the user?

File: pkg/auth/authorizer/abac/abac.go (hunk 1 above)
Comment (erictune) [inline pkg/auth/authorizer/abac/abac.go]: If you do polling, make the flag be something like `--abac-file-poll-period-seconds` or something like that.

Comment (erictune) [inline pkg/auth/authorizer/abac/abac.go]: a default value of 0 would mean don't poll

**Comment by huang195** [comment]
@liggitt You're right. It just reads the files and hope for the best (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/config/file.go#L134).

As a policy file is usually pretty small, maybe this is a reasonable approach?

File: pkg/auth/authorizer/abac/abac.go (hunk 1 above)
Comment (huang195) [inline pkg/auth/authorizer/abac/abac.go]: yep, that sounds reasonable. I'm thinking getting rid of fsnotify in favor of a simpler polling mechanism.

File: pkg/auth/authorizer/abac/abac.go (hunk 3 above)
Comment (huang195) [inline pkg/auth/authorizer/abac/abac.go]: Yeah, I agree, fsnotify just over complicates things here. 

> You still need to handle the case of an incorrect file. Should you crash, as is currently the case when an incorrect file is provided? Or keep using the old one? How to surface this error to the user?

I don't think we should crash when there's an error due to policy reloading as it could put a running cluster to a halt. Keep using the old one but throw out an error (glog.Errorf(...)) is probably a nicer way of keeping the cluster running while informing the admin who's making the change that there's an error to be corrected. On a successful load, we should also do a glog.Infof(...) so the admin can verify that the change has taken effect. The verification step of checking the log should be documented. Is this reasonable?

Comment (erictune) [inline pkg/auth/authorizer/abac/abac.go]: I forgot that the ABAC file is in JSONL format, not JSON format.  So, what I said is not correct.  You cannot detect the difference between incomplete and complete in all cases.

**Comment by erictune** [comment]
I would not rely on the policy file being small.  It might be more than one block, and in theory you can have a break in writing at any block in the file.

**Comment by huang195** [comment]
How about converting the policy file from JSONL to JSON format, so incomplete writes can be detected?

**Comment by huang195** [comment]
Or how about adding a special marker in the last line of the file, e.g., an empty `{}`

**Comment by liggitt** [comment]
I'd definitely prefer a json object over a marker row, but I'm really hesitant to keep adding more complexity to the abac file parse when it's not something we see as a primary method for live-updating authz in the future

**Comment by huang195** [comment]
@liggitt right, that's why I was saying maybe a simple EOF marker would be sufficient. The ABAC code today is already skipping lines starting with `#`, so we could have something like `#EOF#` to mark the end of a file without changing any syntax, and it's backward compatible. 

Moreover, during the initial loading of the policy file, we can detect if `#EOF#` is present. If so, we will start polling for changes; otherwise, we don't. This gets rid of the need of adding a new flag to specify whether or not we should watch for changes, further simplifying things. The only thing missing is the poll frequency, for which, I think we can set to a constant value that's reasonable, e.g., every 5 seconds.

**Comment by huang195** [comment]
@erictune @liggitt PTAL. I modified this PR to use a poll mechanism to detect for file changes. This seems to be a lot more reliable and intuitive than fsnotify. We will detect if a special marker line "### EOF ###" is added to the last line of the policy file, and its presence indicates that we should watch for policy changes and reload if necessary. It will also help prevent reading partial file due to the current use of JSONL format.

Success/Failure messages are printed to the apiserver log to indicate to the admin making the change whether or not a change has taken effect. On an error, we will continue to use the last working policy. This continues to be backward compatible, i.e., if the special marker line is not detected, we will not watch and reload. 

To sum it up, in the spirit of keeping ABAC simple, the only function added in this PR is to watch for file changes and reload if necessary. The only user-perceived difference is if he/she wants a policy file to be continuously monitored, a special marker line needs to be appended to the bottom of the policy file. Otherwise, keep the policy file as is, and it will not be monitored.

File: pkg/auth/authorizer/abac/abac.go [hunk 5]
```diff
@@ -51,13 +60,74 @@ func (p policyLoadError) Error() string {
 
 type policyList []*api.Policy
 
+type ABACPolicy struct {
```
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: no need to export this, I don't think

File: pkg/auth/authorizer/abac/abac.go [hunk 6]
```diff
@@ -51,13 +60,74 @@ func (p policyLoadError) Error() string {
...
+	}
+
+	if eof {
+		go wait.Forever(func() {
+			var lastModifiedTime *time.Time
+			for {
+				time.Sleep(filePollPeriod)
+
+				done := make(chan error)
+				go func() {
```
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: not seeing why this needs to be async

File: pkg/auth/authorizer/abac/abac.go [hunk 7]
```diff
@@ -51,13 +60,74 @@ func (p policyLoadError) Error() string {
...
+					info, err := os.Stat(path)
+					if err != nil {
+						done <- err
+					} else {
+						modTime := info.ModTime()
+						if lastModifiedTime == nil || *lastModifiedTime == modTime {
```
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: skipping on nil `lastModifiedTime` means an update immediately after starting the server, but before the first poll, would be ignored

File: pkg/auth/authorizer/abac/abac.go [hunk 8]
```diff
@@ -215,7 +290,12 @@ func resourceMatches(p api.Policy, a authorizer.Attributes) bool {
 }
 
 // Authorizer implements authorizer.Authorize
-func (pl policyList) Authorize(a authorizer.Attributes) error {
+func (apolicy *ABACPolicy) Authorize(a authorizer.Attributes) error {
+	pl, ok := apolicy.Auth.Load().(policyList)
```
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: this is going to hit every request... I don't know how Load() performs. Seems like we should only do this in cases where we're supporting reload (which I expect to be a vanishingly small set)

File: pkg/auth/authorizer/abac/abac.go (hunk 5 above)
Comment (huang195) [inline pkg/auth/authorizer/abac/abac.go]: right, this is the same as policyList, which doesn't need to be exported

File: pkg/auth/authorizer/abac/abac.go (hunk 6 above)
Comment (huang195) [inline pkg/auth/authorizer/abac/abac.go]: This doesn't need to be async, but it keeps the code inside of `go func() {...` cleaner, otherwise, for each error condition, we would need to `glog.Warning(...)` and `time.Sleep(failureRetryPeriod)`.

Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: I'd rather see it factored out as a reloadIfNeeded() func that returns an error

**Comment by liggitt** [comment]
before we iterate on this too much more, I'd like @erictune's take... this eof marker feels pretty magical, and I'm not sure I'd want to support it long-term.

File: pkg/auth/authorizer/abac/abac.go (hunk 7 above)
Comment (huang195) [inline pkg/auth/authorizer/abac/abac.go]: yep, will catch this condition as well

**Comment by huang195** [comment]
Yep, agree, a bit hacky here. As @erictune suggested, I'm going to carry the other PR (#23502) that implements an etcd-based ABAC authz plugin in my own fork till the openshift authz can be merged into release-1.3. #23502 satisfies most of our requirements, so we're good for now.

However, from now till openshift authz is merged, policy is pretty much static unless one is willing to restart the apiserver (multiple in case of HA setup). I thought it might be beneficial to add this function as a simpler stop-gap solution than my other PR. Maybe we maintain this until 1.3 is out and then revert it to get rid of the magic trick :)

**Comment by erictune** [comment]
I don't like the EOF thing because it is not backward compatible.
I would accept a change that caused the file to be reloaded when SIGHUP is received by the apiserver process.    Assuming @lavalamp can live with a SIGHUP handler in API server.
Presumably you would just import https://golang.org/pkg/os/signal/, add a `signal.Notify(SUGHUP)` to the constructor to ABAC, and add a goroutine that watched the channel, and atomically reloaded.

**Comment by erictune** [comment]
@lavalamp said he was okay with this.  If you want to reopen this PR or send a new one, I will look favorably on it.  We could use the same approach for password and token files for consistency (you don't have to do that, just saying).

**Comment by huang195** [comment]
@erictune I actually a version of this PR using SIGHUP :-) Let me take a look and do a bit of cleanup before submitting it. However, the reason I don't like using signals for this is when you have multiple apiservers in an HA setup, you have to do this on all the apiservers. If one didn't get the signal for whatever reason, you get into this nasty inconsistent state.

**Comment by lavalamp** [comment]
Yeah, HA + files on disk is not a recipe for a good time.

**Comment by erictune** [comment]
If you are happy with maintaining your etcd-based fork for now, then no need to send the SIGHUP PR.  I'll take the time I would have spent reviewing that and spend it on getting the Openshift auth upstreamed.

**Comment by huang195** [comment]
ok, looking forward to seeing the openshift auth merged. I'll leave this as is for now as file-based authz is not really for HA setup, so we're not really gaining much here. Thanks for the help on a couple of related issues/PRs, @erictune and @liggitt.
